native/dtgsja.go

```diff
@@ -0,0 +1,357 @@
+//  jobU == lapack.GSVDNone     Do not compute orthogonal matrix.
+// The behavior is the same for jobV and jobQ with the exception that instead of
+// lapack.GSVDU these accept lapack.GSVDV and lapack.GSVDQ respectively.
+// The matrices U, V and Q must be m×m, p×p and n×n respectively unless the
+// relevant job parameter is lapack.GSVDNone.
+//
+// k and l specify the sub-blocks in the input matrices A and B:
+//  A23 = A[k:min(k+l,m), n-l:n) and B13 = B[0:l, n-l:n]
+// of A and B, whose GSVD is going to be computed by Dtgsja.
+//
+// tola and tolb are the convergence criteria for the Jacobi-Kogbetliantz
+// iteration procedure. Generally, they are the same as used in the preprocessing
+// step, for example,
+//  tola = max(m, n)*norm(A)*eps,
+//  tolb = max(p, n)*norm(B)*eps.
+// Where eps is the machine epsilon.
+//
+// work must have length at least 2*n, otherwise Dtgsja will panic.
+//
+// alpha and beta must have length n or Dtgsja will panic. On exit, alpha and
+// beta contain the generalized singular value pairs of A and B
+//   alpha[0:k] = 1,
+//   beta[0:k]  = 0,
+// if m-k-l >= 0,
+//   alpha[k:k+l] = diag(C),
+//   beta[k:k+l]  = diag(S),
+// if m-k-l < 0,
+//   alpha[k:m]= C, alpha[m:k+l]= 0
+//   beta[k:m] = S, beta[m:k+l] = 1.
+// if k+l < n,
+//   alpha[k+l:n] = 0 and
+//   beta[k+l:n]  = 0.
+//
+// On exit, A[n-k:n, 0:min(k+l,m)] contains the triangular matrix R or part of R
+// and if necessary, B[m-k:l, n+m-k-l:n] contains a part of R.
+//
+// Dtgsja returns whether the routine converged and the number of iteration cycles
+// that were run.
+//
+// Dtgsja is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dtgsja(jobU, jobV, jobQ lapack.GSVDJob, m, p, n, k, l int, a []float64, lda int, b []float64, ldb int, tola, tolb float64, alpha, beta, u []float64, ldu int, v []float64, ldv int, q []float64, ldq int, work []float64) (cycles int, ok bool) {
+	const maxit = 40
+
+	checkMatrix(m, n, a, lda)
+	checkMatrix(p, n, b, ldb)
+
+	if len(alpha) != n {
+		panic(badAlpha)
+	}
+	if len(beta) != n {
+		panic(badBeta)
+	}
+
+	initu := jobU == lapack.GSVDUnit
+	wantu := initu || jobU == lapack.GSVDU
+	if !initu && !wantu && jobU != lapack.GSVDNone {
+		panic(badGSVDJob + "U")
+	}
+	if jobU != lapack.GSVDNone {
+		checkMatrix(m, m, u, ldu)
+	}
+
+	initv := jobV == lapack.GSVDUnit
+	wantv := initv || jobV == lapack.GSVDV
+	if !initu && !wantu && jobU != lapack.GSVDNone {
+		panic(badGSVDJob + "V")
+	}
+	if jobV != lapack.GSVDNone {
+		checkMatrix(p, p, v, ldv)
+	}
+
+	initq := jobQ == lapack.GSVDUnit
+	wantq := initq || jobQ == lapack.GSVDQ
+	if !initu && !wantu && jobU != lapack.GSVDNone {
+		panic(badGSVDJob + "Q")
+	}
+	if jobQ != lapack.GSVDNone {
+		checkMatrix(n, n, q, ldq)
+	}
+
+	if len(work) < 2*n {
+		panic(badWork)
+	}
+
+	// Initialize U, V and Q, if necessary
+	if initu {
+		impl.Dlaset(blas.All, m, m, 0, 1, u, ldu)
+	}
+	if initv {
+		impl.Dlaset(blas.All, m, m, 0, 1, v, ldv)
```

Comment:
Sorry, I haven't read all the discussion, but these are different than the lapack code.